### ISSUE TITLE
Pipenv fails to install dependencies for pytest on CircleCI

### ISSUE THREAD

**Issue by maxalbert** [pypa/pipenv#1645]
I ran into this issue on CircleCI in a private repo which runs tests using `pytest`, and where `pytest` and the other dependencies are installed using `pipenv`.

Until recently this worked fine and the CircleCI build passed. However, after a push to this repo today the CirlcleCI build now fails to install the dependencies for `pytest` - when trying to run `pytest` it complains about the missing module `six`.

Strangely, I can't reproduce this issue on a local machine, but it definitely fails on CircleCI (see link to minimal repo reproducing the error below). I'm not sure if it's specific to CircleCI or simply due to some differences in the setup.

##### Describe your environment

Here is the CirleCI build environment:

1. OS Type: Debian GNU/Linux 8
2. Python version: 3.6.4
3. Pipenv version: 11.1.11

##### Expected result

```
$ pipenv run pytest --version
This is pytest version 3.4.2, imported from /home/circleci/.local/share/virtualenvs/project-zxI9dQ-Q/lib/python3.6/site-packages/pytest.py
```

##### Actual result

```
$ pipenv run pytest --version
Traceback (most recent call last):
  File "/home/circleci/.local/share/virtualenvs/project-zxI9dQ-Q/bin/pytest", line 7, in <module>
    from pytest import main
  File "/home/circleci/.local/share/virtualenvs/project-zxI9dQ-Q/lib/python3.6/site-packages/pytest.py", line 9, in <module>
    from _pytest.config import (
  File "/home/circleci/.local/share/virtualenvs/project-zxI9dQ-Q/lib/python3.6/site-packages/_pytest/config.py", line 9, in <module>
    import six
ModuleNotFoundError: No module named 'six'
```

##### Steps to replicate

Here is a [minimal repo](https://github.com/maxalbert/pipenv_bug) which reproduces the failure.
The output of a failing CircleCI test run can be seen [here](https://circleci.com/gh/maxalbert/pipenv_bug/15).

The CircleCI [config file](https://github.com/maxalbert/pipenv_bug/blob/master/.circleci/config.yml) in this repo runs the following steps:
- Install `pipenv` and `virtualenv`
- Run `pipenv install`, which should install `pytest` and its dependencies
- Run `pytest --version` to verify that `pytest` was installed correctly and can be run (this is the step that fails)

This is the Pipfile used:
```
[[source]]

url = "https://pypi.python.org/simple"
verify_ssl = true
name = "pypi"


[packages]

pytest = "==3.4.2"


[requires]

python_version = "3.6"
```

And here is the `Pipfile.lock` which is generated by `pipenv install`:
```
{
    "_meta": {
        "hash": {
            "sha256": "748122aab9868a7d99e975aaf09590e4f055778cb44e01aa11039d8d9bc2d2bf"
        },
        "pipfile-spec": 6,
        "requires": {
            "python_version": "3.6"
        },
        "sources": [
            {
                "name": "pypi",
                "url": "https://pypi.python.org/simple",
                "verify_ssl": true
            }
        ]
    },
    "default": {
        "pytest": {
            "hashes": [
                "sha256:062027955bccbc04d2fcd5d79690947e018ba31abe4c90b2c6721abec734261b",
                "sha256:117bad36c1a787e1a8a659df35de53ba05f9f3398fb9e4ac17e80ad5903eb8c5"
            ],
            "version": "==3.4.2"
        }
    },
    "develop": {}
}
```

**Comment by kennethreitz** [comment]
you might have to add a `pipenv run pip install six` in there.

**Comment by kennethreitz** [comment]
working on a fix

**Comment by kennethreitz** [comment]
try v11.2.0

**Comment by maxalbert** [comment]
@kennethreitz Many thanks for the super fast reply!

The first thing I tried was indeed `pipenv install six` (out of interest, why did you suggest to run `pip` through `pipenv` instead?). However it then failed with additional missing dependencies of `pytest`, so it didn't seem to be a viable approach.

Re v11.2.0, unfortunately it's still [failing](https://circleci.com/gh/maxalbert/pipenv_bug/16).

**Comment by kennethreitz** [comment]
Are you using `pipenv run pytest`?

**Comment by maxalbert** [comment]
> Are you using pipenv run pytest?

Yes. See last line in the [.circleci/config.yml](https://github.com/maxalbert/pipenv_bug/blob/master/.circleci/config.yml) file.

FYI, I just did a quick bisection and it turns out that this worked fine with pipenv [v11.1.5](https://circleci.com/gh/maxalbert/pipenv_bug/22) but started breaking in [v11.1.6](https://circleci.com/gh/maxalbert/pipenv_bug/21). Maybe this helps track down the issue?

**Comment by kennethreitz** [comment]
"Pipfile.lock not found, creating…"

I suspect this may be part of the problem.

**Comment by maxalbert** [comment]
Hmm, interesting. Is the `Pipfile.lock` file required? Why would `pipenv` not be able to install the dependencies from just the Pipfile alone?

In the original (private) repo where I first encountered this issue I first had the `Pipfile.lock` included but then ran into cross-platform issues between Linux and Mac OS X so had to remove it again. Is there a good workaround for this?

**Comment by kennethreitz** [comment]
It's not required, I think that just may be the source of this bug. 

If you want to help debug, can you try to add the lockfile to your repo and see if that changes behavior?

**Comment by kennethreitz** [comment]
We've massively improved cross-platform compatibility with lockfiles over the past few iteration cycles.

**Comment by dmtucker** [comment]
This does not reproduce for me (though, I have to override python):
```
dtucker@dtucker-wkstn:~/Projects/pipenv_bug $ pipenv --version
pipenv, version 11.2.0
dtucker@dtucker-wkstn:~/Projects/pipenv_bug $ pipenv install
Warning: Python 3.6 was not found on your system…
You can specify specific versions of Python with:
  $ pipenv --python path/to/python
[Fri Mar  9 14:26:41 PST 2018] 1
dtucker@dtucker-wkstn:~/Projects/pipenv_bug $ pipenv --python python3 install
Creating a virtualenv for this project…
Using /usr/bin/python3 to create virtualenv…
⠋Already using interpreter /usr/bin/python3
Using base prefix '/usr'
New python executable in /home/dtucker/.local/share/virtualenvs/pipenv_bug-HDpInDZX/bin/python3
Also creating executable in /home/dtucker/.local/share/virtualenvs/pipenv_bug-HDpInDZX/bin/python
Installing setuptools, pip, wheel...done.

Virtualenv location: /home/dtucker/.local/share/virtualenvs/pipenv_bug-HDpInDZX
Warning: Your Pipfile requires python_version 3.6, but you are using 3.5.3 (/home/d/.local/share/v/p/bin/python).
  $ pipenv check will surely fail.
Warning: Your Pipfile requires python_version 3.6, but you are using 3.5.3 (/home/d/.local/share/v/p/bin/python).
  $ pipenv check will surely fail.
Pipfile.lock not found, creating…
Locking [dev-packages] dependencies…
Locking [packages] dependencies…
Updated Pipfile.lock (c2d2bf)!
Installing dependencies from Pipfile.lock (c2d2bf)…
  🐍   ▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉ 5/5 — 00:00:01
To activate this project's virtualenv, run the following:
 $ pipenv shell
dtucker@dtucker-wkstn:~/Projects/pipenv_bug $ pipenv run pytest --version
Warning: Your Pipfile requires python_version 3.6, but you are using 3.5.3 (/home/d/.local/share/v/p/bin/python).
  $ pipenv check will surely fail.
This is pytest version 3.4.2, imported from /home/dtucker/.local/share/virtualenvs/pipenv_bug-HDpInDZX/lib/python3.5/site-packages/pytest.py
```
Maybe try `pip install --user` instead of `sudo pip install`?

**Comment by kennethreitz** [comment]
That's basically lead to the bugs you're seeing, involved massive overhauls to our architecture :)

**Comment by dmtucker** [comment]
> Ignoring pluggy: markers 'python_version == ">=2.7,!=3.0.*,!=3.1.*,!=3.2.*,!=3.3.*"' don't match your environment
Ignoring py: markers 'python_version == ">=2.7,!=3.0.*,!=3.1.*,!=3.2.*,!=3.3.*"' don't match your environment

I was hitting these earlier this morning actually. Upgrading to 11.2 seemed to fix my issue... :thinking:

**Comment by maxalbert** [comment]
@dmtucker It also fails if I install pipenv via `pip install --user`. However, interestingly in this case it's complaining about the missing dependency `py` instead of `six` (see comparison between the CircleCI runs with pipenv installed [via sudo](https://circleci.com/gh/maxalbert/pipenv_bug/28) and with the [--user switch](https://circleci.com/gh/maxalbert/pipenv_bug/27)).

**Comment by dmtucker** [comment]
Oh looks like [that run](https://circleci.com/gh/maxalbert/pipenv_bug/27) had a Pipfile.lock? You'll need to recreate it to get around those.

**Comment by maxalbert** [comment]
Argh sorry, you are right. I logged into CircleCI via ssh to check the behaviour with `--user` and it must have re-used the Pipfile.lock from the actual CI run...

**Comment by maxalbert** [comment]
@kennethreitz @dmtucker And indeed, if I add the Pipfile.lock to the repo the issue disappears (see [this](https://circleci.com/gh/maxalbert/pipenv_bug/29) run).

So my understanding is that this is still a bug because it should be possible to use the Pipfile alone, but it can be worked around by committing the Pipfile.lock (and hopefully cross-platform issues may have been fixed in the meantime)?

**Comment by dmtucker** [comment]
A lock file shouldn't be required... Generally speaking nothing good will come from `sudo pip`ing.
Just to confirm, run 27 did have a Pipfile.lock or no?

**Comment by kennethreitz** [comment]
@maxalbert correct.

**Comment by kennethreitz** [comment]
can you try with latest and see if this still happens?

**Comment by jtratner** [comment]
Still happening, just tried with the latest.

```
TDIR=$(mktemp -d)
cd $TDIR
pipenv install pytest
pipenv --rm
pipenv install
pipenv run pytest -h
```

ends up only installing one package and this traceback

```
Traceback (most recent call last):
  File "/Users/jtratner/.virtualenvs/tmp.JmWiHJvy-6om4Ua26/bin/pytest", line 7, in <module>
    from pytest import main
  File "/Users/jtratner/.virtualenvs/tmp.JmWiHJvy-6om4Ua26/lib/python2.7/site-packages/pytest.py", line 9, in <module>
    from _pytest.config import (
  File "/Users/jtratner/.virtualenvs/tmp.JmWiHJvy-6om4Ua26/lib/python2.7/site-packages/_pytest/config.py", line 9, in <module>
    import six
ImportError: No module named six
```

**Comment by kennethreitz** [comment]
And the contents of the lockfile?

**Comment by jtratner** [comment]
@kennethreitz 

```
{
    "_meta": {
        "hash": {
            "sha256": "497286c97fbfe422be45e95a2ef6cf044f88a7d0c777c0a926674894ca547f9b"
        },
        "host-environment-markers": {
            "implementation_name": "cpython",
            "implementation_version": "3.6.1",
            "os_name": "posix",
            "platform_machine": "x86_64",
            "platform_python_implementation": "CPython",
            "platform_release": "16.7.0",
            "platform_system": "Darwin",
            "platform_version": "Darwin Kernel Version 16.7.0: Thu Jan 11 22:59:40 PST 2018; root:xnu-3789.73.8~1/RELEASE_X86_64",
            "python_full_version": "3.6.1",
            "python_version": "3.6",
            "sys_platform": "darwin"
        },
        "pipfile-spec": 6,
        "requires": {},
        "sources": [
            {
                "name": "pypi",
                "url": "https://pypi.python.org/simple",
                "verify_ssl": true
            }
        ]
    },
    "default": {
        "pytest": {
            "hashes": [
                "sha256:062027955bccbc04d2fcd5d79690947e018ba31abe4c90b2c6721abec734261b",
                "sha256:117bad36c1a787e1a8a659df35de53ba05f9f3398fb9e4ac17e80ad5903eb8c5"
            ],
            "version": "==3.4.2"
        }
    },
    "develop": {}
}
```

**Comment by kennethreitz** [comment]
That’s not latest. We removed host environment markers from the lockfile several iterations ago.

**Comment by jtratner** [comment]
my bad - I'd installed an older version of pipenv to experiment with something, here's what I get with master on an editable:

```
{
    "_meta": {
        "hash": {
            "sha256": "d9a5e8e7a8376e93dc640e253fc91b5a8259bd29e7286ec7d740919a6ea52cde"
        },
        "pipfile-spec": 6,
        "requires": {
            "python_version": "2.7"
        },
        "sources": [
            {
                "name": "pypi",
                "url": "https://pypi.python.org/simple",
                "verify_ssl": true
            }
        ]
    },
    "default": {
        "pytest": {
            "hashes": [
                "sha256:062027955bccbc04d2fcd5d79690947e018ba31abe4c90b2c6721abec734261b",
                "sha256:117bad36c1a787e1a8a659df35de53ba05f9f3398fb9e4ac17e80ad5903eb8c5"
            ],
            "version": "==3.4.2"
        }
    },
    "develop": {}
}
```

```
pipenv, version 11.2.2
```

**Comment by kennethreitz** [comment]
Still the error?

**Comment by jtratner** [comment]
@kennethreitz - yes - put up a gist just to clarify that I'm really truly working off current master :) https://gist.github.com/jtratner/440e949b64889aa5afe90e1c871385af

**Comment by maxalbert** [comment]
@kennethreitz The issue appears to be fixed with pipenv v11.2.2 in the minimal repo to which I linked above. However, the error still appears in my private repo. :(

Very strangely, even in the minimal repo it failed at first (with a `ValueError: Your dependencies could not be resolved.`). But then after I added a commit and reverted it again the resulting CirlecCI build passed (for reference, here are the two corresponding outputs: [failing](https://circleci.com/gh/maxalbert/pipenv_bug/30) and [passing](https://circleci.com/gh/maxalbert/pipenv_bug/32)).

So given that it still fails in my private repo and @jtratner also reports the same failure there still appears to be a bug somewhere.

FYI I won't have time to debug this further until mid next week, but I hope the information provided so far can help shed some light on the issue. Many thanks again for your help so far!

**Comment by kennethreitz** [comment]
try 11.3.3

**Comment by maxalbert** [comment]
Yup, it seems to be fixed in the latest version (tested with 11.4.0). Many thanks again for your support and responsiveness!

**Comment by kennethreitz** [comment]
Very sorry for the bugs. We got our shit together now :)
